### PR TITLE
DYN-6901 Pm - publish process improvements 

### PR DESCRIPTION
**Pull request by dnenov**
### Purpose

This is the first PR of a series of changes to the `publish a package using retain folder structure` which aims to fix and updates a lot of the issues the process has had since its launch. I tried to keep the separate PRs relevant to a single change and as small as possible. 

### Changes in this PR
- the main change is to the `BindParentToChild` and `GetCommonPaths` methods. Historically, after loading files during the publish package process, **Dynamo did not capture the relationships between files and folders**, as the only option was to cluster everything into the 4 main folders - `bin`, `doc`, `dyf`, `extra`. 
After introducing the concept of retaining the existing folder structure, this information became important. The 2 methods above do the job by post-processing all the 'raw' file paths each time after a new file or folder has been added to the collection via the UI. 
- added test coverage for the above methods
- support for capitalized file extensions - '.DYF', etc, (`x.EndsWith(".dyf")` -> `x.ToLower().EndsWith(".dyf")`)
- highlights existing limitation for further discussion

### Limitations to the current process
- `Dynamo does not allow loading of multiple .dll versions` - created a test to highlight this limitation: EnsureMultipleVersionsOfAssembly_CannotBeLoaded()

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [x] This PR modifies some build requirements and the readme is updated
- [x] This PR contains no files larger than 50 MB

### Release Notes

- rework on the logic around building file and folder relationships that supports the 'retain folder structure' workflow
- support for capitalized file extensions (dyf, dll, etc)
- test coverage

### Reviewers

@QilongTang 
@zeusongit 
@reddyashish 

### FYIs

@Amoursol

### CODE REVIEW

**Comment by reddyashish** [comment]
@dnenov Do we have a task for this?

**Comment by dnenov** [comment]
> @dnenov Do we have a task for this?

Hi @reddyashish - it is part of the PackageManager update, no jira task yet, and it's still WIP

**Comment by QilongTang** [comment]
> > @dnenov Do we have a task for this?
> 
> Hi @reddyashish - it is part of the PackageManager update, no jira task yet, and it's still WIP

Hi @dnenov is it not logged as part of release epic https://jira.autodesk.com/browse/DYN-5270?

File: test/DynamoCoreWpfTests/PublishPackageViewModelTests.cs [hunk 1]
```diff
@@ -300,5 +300,195 @@ public void AssertAddChildRecursively_IsSuccessful()
...
+
+            //arrange the second assembly version
+            packagedirectory = Path.Combine(TestDirectory, "pkgs\\PackageManager\\2.0");
+            var version_2 = Directory.GetFiles(packagedirectory);
+            var secondAssembly = version_2.First();
+
+            //now add the second assembly version file
+            vm.AddFile(secondAssembly);
+
+            //TODO: assert - do we expect to see 1 or 2 here?
```
Comment (QilongTang) [inline test/DynamoCoreWpfTests/PublishPackageViewModelTests.cs:329]: Nice, this is a good one to revisit later.

File: test/DynamoCoreWpfTests/PublishPackageViewModelTests.cs [hunk 2]
```diff
@@ -300,5 +300,195 @@ public void AssertAddChildRecursively_IsSuccessful()
...
+
+
+        #region CreateContentsRelationships
+
+        [Test]
+        public void CreatesCorrectRelationships_ControlTest()
+        {
+            var vm = new PublishPackageViewModel(ViewModel);
+            var items = new Dictionary<string, PackageItemRootViewModel>();
+            var files = new[] { new FileInfo(@"C:\pkg\file1.dyn"), new FileInfo(@"C:\pkg\file2.DYN") };
```
Comment (QilongTang) [inline test/DynamoCoreWpfTests/PublishPackageViewModelTests.cs:341]: Nice case with capitalized file format

**Review by QilongTang — APPROVED** [review]

**Comment by QilongTang** [comment]
@dnenov Can you take a look at the reported regression [DynamoCoreWpfTests.PublishPackageViewModelTests.EnsureMultipleVersionsOfAssembly_CannotBeLoaded](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16240/testReport/junit/DynamoCoreWpfTests/PublishPackageViewModelTests/EnsureMultipleVersionsOfAssembly_CannotBeLoaded/) to see if it needs to be updated?

**Comment by dnenov** [comment]
> @dnenov Can you take a look at the reported regression [DynamoCoreWpfTests.PublishPackageViewModelTests.EnsureMultipleVersionsOfAssembly_CannotBeLoaded](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16240/testReport/junit/DynamoCoreWpfTests/PublishPackageViewModelTests/EnsureMultipleVersionsOfAssembly_CannotBeLoaded/) to see if it needs to be updated?

Yes, I think it was the test files being ignored. I added them back in, it should pass now I hope.

**Comment by QilongTang** [comment]
Passed at https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16263/parameters/

**Pull request merged**